### DISCUSSION TITLE
How to use distance3d.containment_test.points_in_box?

### DISCUSSION THREAD

**Discussion by Puiching-Memory** [AlexanderFabisch/distance3d#89] [Q&A] [answered]
I just learned about this python library and it's great!
I now have some 3D boxes and some 3D points, and I want to eliminate all the 3D points that are not in the box.
It seems that this function can help me, but I'm not sure what box2origin means here.
I know this is a 3*3 rotation matrix +1*3 translation matrix, maybe a transformation from the origin to the current 3D frame?

**Comment by AlexanderFabisch** [comment, marked as answer]
Hi, as the docstring says, it's a 4x4 numpy array. It is organized as follows:

```
[[ R11, R12, R13, t1],
  [ R21, R22, R23, t2],
  [ R31, R32, R33, t3],
  [ 0, 0, 0, 1]]
```

All components with R belong to the rotation matrix and all components with t to the translation vector.

box2origin represents the pose of the box center frame in the origin frame. This is the same as an active transformation of points from box frame to the origin frame.

The function will do exactly what you want to do.

**Comment by Puiching-Memory** [comment]
I made an example that took 0.011s to calculate 100000 points, very good results!
```python
import numpy as np
from scipy.spatial.transform import Rotation as R
import distance3d.containment_test

# Example
from pyinstrument import Profiler
profiler = Profiler()
profiler.start()


point_cloud = np.random.randn(100000, 3) * 2  # rand point
box_center = np.array([1, 1, 1])  # center point
box_size = np.array([2, 2, 2])  # size
box_rotation = R.from_euler('xyz', [0, 0, 45], degrees=True).as_matrix()  # Rotation Z 45deg
box_matrix = np.vstack((np.hstack((box_rotation, box_center.reshape(3, 1))), np.array([0, 0, 0, 1])))
print(box_matrix)

# fitter
points_in_box = distance3d.containment_test.points_in_box(point_cloud, box_matrix, box_size)
point_cloud = point_cloud[points_in_box]
print(point_cloud.shape)


profiler.stop()
profiler.print()

with open("profiler.html", "w") as f:
    f.write(profiler.output_html())

import matplotlib.pyplot as plt

fig = plt.figure()
ax = fig.add_subplot(111, projection='3d')
ax.scatter(point_cloud[:,0], point_cloud[:,1], point_cloud[:,2], c='r', marker='o')
ax.set_xlabel('X Label')
ax.set_ylabel('Y Label')
ax.set_zlabel('Z Label')
plt.show()


```
![image](https://github.com/user-attachments/assets/22bc6aff-66e2-43f5-b6a5-3439dc81de39)

```
[[ 0.70710678 -0.70710678  0.          1.        ]
 [ 0.70710678  0.70710678  0.          1.        ]
 [ 0.          0.          1.          1.        ]
 [ 0.          0.          0.          1.        ]]
```

```
0.011 <module>  _test_code.py:1
├─ 0.005 points_in_box  distance3d\containment_test.py:206
│     [3 frames hidden]  numpy, <built-in>
│        0.003 [self]  distance3d\containment_test.py
├─ 0.004 RandomState.randn  <built-in>
├─ 0.001 print  <built-in>
└─ 0.001 _array_str_implementation  numpy\core\arrayprint.py:1595
      [3 frames hidden]  numpy
```

**Reply by AlexanderFabisch** [reply]
Glad that you found it to be useful in your case. I was also thinking about accelerating the code further with JAX. There is an example for the capsule containment test on this branch: https://github.com/AlexanderFabisch/distance3d/tree/exp/jax . For the first function call there would be a bit more overhead to compile it, but then you would be able to run code in parallel on the GPU. Not sure how much it would accelerate the box containment test though.
